src/query/graphite/storage/m3_wrapper.go

```diff
@@ -134,17 +134,18 @@ func truncateBoundsToResolution(
 	end time.Time,
 	resolution time.Duration,
 ) (time.Time, time.Time) {
-	truncatedStart := start.Truncate(resolution)
-	// NB: if truncated start matches start, it's already valid.
-	if truncatedStart.Before(start) {
-		start = truncatedStart.Add(resolution)
-	}
-
-	length := float64(end.Sub(truncatedStart))
+	truncatedEnd := end.Truncate(resolution)
+	length := float64(truncatedEnd.Sub(start))
 	steps := math.Floor(length / float64(resolution))
+
 	truncatedLength := time.Duration(steps) * resolution
-	end = start.Add(truncatedLength)
-	return start, end
+	truncatedStart := truncatedEnd.Add(truncatedLength * -1)
+
+	// start one step earlier, unless the start time was already truncated
+	if !truncatedStart.Equal(start) {
+		truncatedStart = truncatedStart.Add(resolution * -1)
```

Comment:
discussed offline, this special handling seems a bit of intuitive. needs to be changed.

Reply:
Removed